### PR TITLE
OpenSSL 3.0

### PR DESCRIPTION
**Pull request by nicowilliams**

### CODE REVIEW

File: kdc/pkinit-ec.c [hunk 1]
```diff
@@ -69,101 +73,85 @@
...
+        ret = krb5_enomem(context);
+    if (ret == 0 && EVP_PKEY_derive_init(pctx) != 1)
+        ret = krb5_enomem(context);
+    if (ret == 0 &&
+        EVP_PKEY_CTX_set_ecdh_kdf_type(pctx, EVP_PKEY_ECDH_KDF_NONE) != 1)
+        krb5_set_error_message(context, ret = KRB5KRB_ERR_GENERIC,
+                               "Could not generate an ECDH key agreement private key "
+                               "(EVP_PKEY_CTX_set_dh_kdf_type)");
+    if (ret == 0 &&
+        EVP_PKEY_derive_set_peer_ex(pctx, ec_key_pub, 1) == 0)
```
Comment (jaltman) [inline kdc/pkinit-ec.c:124]: Reading the EVP_PKEY_derive_set_peer_ex() sources it appears that 0 is success.   Failure codes are negative values.

File: kdc/pkinit-ec.c [hunk 2]
```diff
@@ -69,101 +73,85 @@
...
+        krb5_set_error_message(context, ret = KRB5KRB_ERR_GENERIC,
+                               "Could not generate an ECDH key agreement private key "
+                               "(EVP_PKEY_derive_set_peer_ex)");
+    if (ret == 0 &&
+        (EVP_PKEY_derive(pctx, NULL, &size) == 0 || size == 0))
```
Comment (jaltman) [inline kdc/pkinit-ec.c:129]: EVP_PKEY_derive appears to return 0 for success.  Errors are negative values.

File: kdc/pkinit-ec.c [hunk 3]
```diff
@@ -69,101 +73,85 @@
...
+        krb5_set_error_message(context, ret = KRB5KRB_ERR_GENERIC,
+                               "Could not generate an ECDH key agreement private key "
+                               "(EVP_PKEY_derive)");
+    if (ret == 0 && (p = malloc(size)) == NULL)
+        ret = krb5_enomem(context);
+    if (ret == 0 &&
+        (EVP_PKEY_derive(pctx, p, &size) == 0 || size == 0))
```
Comment (jaltman) [inline kdc/pkinit-ec.c:136]: EVP_PKEY_derive appears to return 0 for success.

File: lib/hcrypto/evp.c [hunk 4]
```diff
@@ -857,60 +871,72 @@ EVP_CipherUpdate(EVP_CIPHER_CTX *ctx, void *out, int *outlen,
...
+     * If there in no bytes left over from the last Update and the
+     * input length is on a block boundary, then we can take a
+     * shortcut (and preformance gain) and directly encrypt the
+     * data.
      */
-    if (ctx->buf_len == 0 && (inlen & ctx->block_mask) == 0) {
+    if (ctx->buf_len == 0 && inlen && (inlen & ctx->block_mask) == 0) {
+        /*
+         * OpenSSL's EVP_Cipher() returns the number of bytes on success, 0 on
+         * failure.
```
Comment (jaltman) [inline lib/hcrypto/evp.c:883]: EVP_Cipher calls ctx->cipher->do_cipher().   EVP_Cipher cannot be assigned to do_cipher.    do_cipher() functions return 1 for success and any other value is an error.

File: lib/hcrypto/evp.c [hunk 5]
```diff
@@ -857,60 +871,72 @@ EVP_CipherUpdate(EVP_CIPHER_CTX *ctx, void *out, int *outlen,
...
+         */
 	ret = (*ctx->cipher->do_cipher)(ctx, out, in, inlen);
-	if (ret == 1)
+	if (ret == 1 || ret == inlen) {
```
Comment (jaltman) [inline lib/hcrypto/evp.c:886]: I do not believe the "ret == inlen" check is correct.

File: lib/hcrypto/evp.c [hunk 6]
```diff
@@ -857,60 +871,72 @@ EVP_CipherUpdate(EVP_CIPHER_CTX *ctx, void *out, int *outlen,
...
+            return 1;
+        }
+
+        /* Fill the buffer and encrypt */
+        memcpy(ctx->buf + ctx->buf_len, in, left);
+        ret = (*ctx->cipher->do_cipher)(ctx, out, ctx->buf, blocksize);
+        memset_s(ctx->buf, blocksize, 0, blocksize);
+        /*
+         * OpenSSL's EVP_Cipher() returns the number of bytes on success, 0 on
+         * failure.
```
Comment (jaltman) [inline lib/hcrypto/evp.c:912]: do_cipher() cannot be EVP_Cipher().   ctx->cipher->do_cipher() is called from EVP_Cipher().

File: lib/hcrypto/evp.c [hunk 7]
```diff
@@ -857,60 +871,72 @@ EVP_CipherUpdate(EVP_CIPHER_CTX *ctx, void *out, int *outlen,
...
+         */
+        if (ret != 1 && ret != blocksize)
```
Comment (jaltman) [inline lib/hcrypto/evp.c:914]: I do not believe the "ret != blocksize" check is correct.

File: lib/hcrypto/evp.c [hunk 8]
```diff
@@ -857,60 +871,72 @@ EVP_CipherUpdate(EVP_CIPHER_CTX *ctx, void *out, int *outlen,
...
 	ctx->buf_len = (inlen & ctx->block_mask);
 	inlen &= ~ctx->block_mask;
 
-	ret = (*ctx->cipher->do_cipher)(ctx, out, in, inlen);
-	if (ret != 1)
-	    return ret;
+        if (inlen) {
+            /* Encrypt all the whole blocks of input that we have */
+            ret = (*ctx->cipher->do_cipher)(ctx, out, in, inlen);
+            if (ret != 1 && ret != inlen)
```
Comment (jaltman) [inline lib/hcrypto/evp.c:932]: I do not believe the "ret != inlen" check is correct.    do_cipher() functions return 1 on success and any other value is an error.

File: lib/hcrypto/evp.c [hunk 9]
```diff
@@ -952,7 +978,11 @@ EVP_CipherFinal_ex(EVP_CIPHER_CTX *ctx, void *out, int *outlen)
 	memset(ctx->buf + ctx->buf_len, 0, left);
 	ret = (*ctx->cipher->do_cipher)(ctx, out, ctx->buf, blocksize);
 	memset_s(ctx->buf, blocksize, 0, blocksize);
-	if (ret != 1)
+        /*
+         * OpenSSL's EVP_Cipher() returns the number of bytes on success, 0 on
+         * failure.
```
Comment (jaltman) [inline lib/hcrypto/evp.c:983]: ctx->cipher->do_cipher() cannot be set to EVP_Cipher().   EVP_Cipher() calls ctx->cipher->do_cipher().

File: lib/hcrypto/evp.c [hunk 10]
```diff
@@ -952,7 +978,11 @@ EVP_CipherFinal_ex(EVP_CIPHER_CTX *ctx, void *out, int *outlen)
...
+         */
+	if (ret != 1 && ret != blocksize)
```
Comment (jaltman) [inline lib/hcrypto/evp.c:985]: I do not believe the "ret != blocksize" is correct.

File: lib/krb5/pkinit-ec.c [hunk 11]
```diff
@@ -172,51 +169,56 @@ _krb5_pk_rd_pa_reply_ecdh_compute_key(krb5_context context,
...
+    if (ret == 0 &&
+        EVP_PKEY_CTX_set_ecdh_kdf_type(pctx, EVP_PKEY_ECDH_KDF_NONE) != 1)
+        ret = krb5_enomem(context);
+    if (ret == 0 &&
+        (public = d2i_PublicKey(EVP_PKEY_EC, &template, &in, in_sz)) == NULL)
+        krb5_set_error_message(context,
+                               ret = HX509_PARSING_KEY_FAILED,
+                               "PKINIT: Can't parse the KDC's ECDH public key");
+    if (ret == 0 &&
+        EVP_PKEY_derive_set_peer_ex(pctx, public, 1) == 0)
```
Comment (jaltman) [inline lib/krb5/pkinit-ec.c:195]: EVP_PKEY_derive_set_peer_ex() returns 0 on success.

File: lib/krb5/pkinit-ec.c [hunk 12]
```diff
@@ -172,51 +169,56 @@ _krb5_pk_rd_pa_reply_ecdh_compute_key(krb5_context context,
...
+        krb5_set_error_message(context,
+                               ret = KRB5KRB_ERR_GENERIC,
+                               "Could not derive ECDH shared secret for PKINIT key exchange "
+                               "(EVP_PKEY_derive_set_peer_ex)");
+    if (ret == 0 &&
+        (EVP_PKEY_derive(pctx, NULL, &shared_len) == 0 || shared_len == 0))
```
Comment (jaltman) [inline lib/krb5/pkinit-ec.c:201]: EVP_PKEY_derive() executes the function set by EVP_PKEY_meth_set_derive().   I believe this function is required to return 0 on success.

File: lib/krb5/pkinit-ec.c [hunk 13]
```diff
@@ -172,51 +169,56 @@ _krb5_pk_rd_pa_reply_ecdh_compute_key(krb5_context context,
...
+                               "Could not derive ECDH shared secret for PKINIT key exchange "
+                               "(EVP_PKEY_derive to get length)");
+    if (ret == 0 && shared_len > INT_MAX)
+        krb5_set_error_message(context,
+                               ret = KRB5KRB_ERR_GENERIC,
+                               "Could not derive ECDH shared secret for PKINIT key exchange "
+                               "(shared key too large)");
+    if (ret == 0 && (*out = malloc(shared_len)) == NULL)
+        ret = krb5_enomem(context);
+    if (ret == 0 && EVP_PKEY_derive(pctx, *out, &shared_len) == 0)
```
Comment (jaltman) [inline lib/krb5/pkinit-ec.c:213]: I believe EVP_PKEY_derive() returns 0 upon success.

**Review by jaltman — CHANGES_REQUESTED** [review]
The OpenSSL functions are very poorly documented.    Reference counting for many interfaces were introduced in 3.0.x

File: lib/krb5/pkinit-ec.c (hunk 13 above)
Comment (nicowilliams) [reply lib/krb5/pkinit-ec.c:213]: image

File: kdc/pkinit-ec.c (hunk 1 above)
Comment (nicowilliams) [reply kdc/pkinit-ec.c:124]:
```
377 int EVP_PKEY_derive_set_peer_ex(EVP_PKEY_CTX *ctx, EVP_PKEY *peer,
378                                 int validate_peer)
379 {
380     int ret = 0, check;
...
454     ret = ctx->pmeth->ctrl(ctx, EVP_PKEY_CTRL_PEER_KEY, 0, peer);
455
456     if (ret <= 0)
457         return ret;
458
459     if (ret == 2)
460         return 1;
...
495     EVP_PKEY_up_ref(peer);
496     return 1;
497 #endif
498 }
```
Line 496 makes it clear that it's 1 for success.  Lines 459-460 must be about `ctx->pmeth->ctrl()` returning 2 in some success case.

There's no docs on this function's return value.  The code works though, and the surrounding `EVP_PKEY_derive*()` functions are documented as returning 1 on success.

File: kdc/pkinit-ec.c (hunk 2 above)
Comment (nicowilliams) [reply kdc/pkinit-ec.c:129]: See above.

File: lib/hcrypto/evp.c (hunk 4 above)
Comment (nicowilliams) [reply lib/hcrypto/evp.c:883]: Ah yes, I _fixed_ that harder in `cipher_do_cipher()`, so I can take this change out.

File: lib/hcrypto/evp.c (hunk 5 above)
Comment (nicowilliams) [reply lib/hcrypto/evp.c:886]: Yes, because I fixed this harder in `cipher_do_cipher()`.

**Comment by nicowilliams** [comment]
So we're going to have to have some ifdef spaghetti to deal with OpenSSL 1.1 vs 3.0.

**Comment by nicowilliams** [comment]
Also, note to self and others for when we finally remove weak ciphers: we're going to have to re-generate PKCS#12 test samples using stronger ciphers, and possibly add new PBEs.

File: lib/krb5/pkinit-ec.c (hunk 13 above)
Comment (jaltman) [reply lib/krb5/pkinit-ec.c:230]: > image

Upon re-reading the OpenSSL 3.0 source, it is consistent.   The call usage in the OpenSSL sources is "EVP_PKEY_derive() <= 0" return error.

File: lib/hcrypto/evp.c [hunk 14]
```diff
@@ -969,7 +970,7 @@ EVP_CipherFinal_ex(EVP_CIPHER_CTX *ctx, void *out, int *outlen)
  * @param in in data to the operation.
  * @param size length of data.
  *
- * @return 1 on success.
+ * @return bytes encrypted on success, zero on failure.
```
Comment (jaltman) [inline lib/hcrypto/evp.c:986]: At line 979 below do_cipher() is going to return 1 for success and any other value (less than or equal to zero) is an error.  Therefore the return value from EVP_Cipher() must not be the return from do_cipher() and instead be either zero or number of bytes.   If we have nothing else, then return 'size' on success.

File: lib/hcrypto/evp-openssl.h [hunk 15]
```diff
@@ -37,25 +37,31 @@
 #define HEIM_EVP_OSSL_H 1
 
 /* symbol renaming */
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
```
Comment (jaltman) [inline lib/hcrypto/evp-openssl.h:40]: HAVE_OPENSSL_30 ?

**Review by jaltman — CHANGES_REQUESTED** [review]

File: lib/hcrypto/evp.c (hunk 14 above)
Comment (nicowilliams) [reply lib/hcrypto/evp.c:986]: The OpenSSL docs say:

```
EVP_CipherFinal_ex() returns 0 for a decryption failure or 1 for success.
```

File: lib/hx509/crypto-ec.c [hunk 16]
```diff
@@ -215,74 +360,130 @@ ecdsa_create_signature(hx509_context context,
...
+    if (ret == 0 && EVP_DigestSignUpdate(mdctx, data->data, data->length) == 0)
+        ret = HX509_CMS_FAILED_CREATE_SIGATURE;
+    if (ret == 0 && EVP_DigestSignFinal(mdctx, NULL, &sig->length) == 0)
+        ret = HX509_CMS_FAILED_CREATE_SIGATURE;
+    if (ret == 0 && (sig->data = malloc(sig->length)) == NULL)
+        ret = hx509_enomem(context);
+    if (ret == 0 && EVP_DigestSignFinal(mdctx, sig->data, &sig->length) == 0)
+        ret = HX509_CMS_FAILED_CREATE_SIGATURE;
+
+    //ret = ECDSA_sign(-1, indata.data, indata.length,
```
Comment (lhoward) [inline lib/hx509/crypto-ec.c:397]: Is this something you want to delete?

File: lib/hx509/crypto-ec.c [hunk 17]
```diff
@@ -215,74 +360,130 @@ ecdsa_create_signature(hx509_context context,
...
+                    sig->data, &siglen, signer->private_key.ecdsa);
     der_free_octet_string(&indata);
     if (ret != 1) {
-	ret = HX509_CMS_FAILED_CREATE_SIGATURE;
-	hx509_set_error_string(context, 0, ret,
-			       "ECDSA sign failed: %d", ret);
-	goto error;
+       ret = HX509_CMS_FAILED_CREATE_SIGATURE;
+       hx509_set_error_string(context, 0, ret,
+                              "ECDSA sign failed: %d", ret);
```
Comment (lhoward) [inline lib/hx509/crypto-ec.c:462]: I guess, reindentation can be nice to avoid but, whatever.

File: lib/hx509/crypto-ec.c (hunk 16 above)
Comment (nicowilliams) [reply lib/hx509/crypto-ec.c:397]: Yes!  That is left over from when I was writing the code.

**Comment by nicowilliams** [comment]
The one build failure is due to check-iprop.

**Pull request merged**